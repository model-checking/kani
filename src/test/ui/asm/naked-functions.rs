// needs-asm-support
// ignore-nvptx64
// ignore-spirv
// ignore-wasm32

#![feature(asm)]
#![feature(llvm_asm)]
#![feature(naked_functions)]
#![feature(or_patterns)]
#![feature(asm_const, asm_sym)]
#![crate_type = "lib"]
#![allow(deprecated)] // llvm_asm!

#[repr(C)]
pub struct P {
    x: u8,
    y: u16,
}

#[naked]
pub unsafe extern "C" fn patterns(
    mut a: u32,
    //~^ ERROR patterns not allowed in naked function parameters
    &b: &i32,
    //~^ ERROR patterns not allowed in naked function parameters
    (None | Some(_)): Option<std::ptr::NonNull<u8>>,
    //~^ ERROR patterns not allowed in naked function parameters
    P { x, y }: P,
    //~^ ERROR patterns not allowed in naked function parameters
) {
    asm!("", options(noreturn))
}

#[naked]
pub unsafe extern "C" fn inc(a: u32) -> u32 {
    //~^ WARN naked functions must contain a single asm block
    //~| WARN this was previously accepted
    a + 1
    //~^ ERROR referencing function parameters is not allowed in naked functions
}

#[naked]
pub unsafe extern "C" fn inc_asm(a: u32) -> u32 {
    asm!("/* {0} */", in(reg) a, options(noreturn));
    //~^ ERROR referencing function parameters is not allowed in naked functions
    //~| WARN only `const` and `sym` operands are supported in naked functions
    //~| WARN this was previously accepted
}

#[naked]
pub unsafe extern "C" fn inc_closure(a: u32) -> u32 {
    //~^ WARN naked functions must contain a single asm block
    //~| WARN this was previously accepted
    (|| a + 1)()
}

#[naked]
pub unsafe extern "C" fn unsupported_operands() {
    //~^ WARN naked functions must contain a single asm block
    //~| WARN this was previously accepted
    let mut a = 0usize;
    let mut b = 0usize;
    let mut c = 0usize;
    let mut d = 0usize;
    let mut e = 0usize;
    const F: usize = 0usize;
    static G: usize = 0usize;
    asm!("/* {0} {1} {2} {3} {4} {5} {6} */",
         //~^ WARN asm in naked functions must use `noreturn` option
         //~| WARN this was previously accepted
         in(reg) a,
         //~^ WARN only `const` and `sym` operands are supported in naked functions
         //~| WARN this was previously accepted
         inlateout(reg) b,
         inout(reg) c,
         lateout(reg) d,
         out(reg) e,
         const F,
         sym G,
    );
}

#[naked]
pub extern "C" fn missing_assembly() {
    //~^ WARN naked functions must contain a single asm block
    //~| WARN this was previously accepted
}

#[naked]
pub extern "C" fn too_many_asm_blocks() {
    //~^ WARN naked functions must contain a single asm block
    //~| WARN this was previously accepted
    asm!("");
    //~^ WARN asm in naked functions must use `noreturn` option
    //~| WARN this was previously accepted
    asm!("");
    //~^ WARN asm in naked functions must use `noreturn` option
    //~| WARN this was previously accepted
    asm!("");
    //~^ WARN asm in naked functions must use `noreturn` option
    //~| WARN this was previously accepted
    asm!("", options(noreturn));
}

pub fn outer(x: u32) -> extern "C" fn(usize) -> usize {
    #[naked]
    pub extern "C" fn inner(y: usize) -> usize {
        //~^ WARN naked functions must contain a single asm block
        //~| WARN this was previously accepted
        *&y
        //~^ ERROR referencing function parameters is not allowed in naked functions
    }
    inner
}

#[naked]
unsafe extern "C" fn llvm() -> ! {
    //~^ WARN naked functions must contain a single asm block
    //~| WARN this was previously accepted
    llvm_asm!("");
    //~^ WARN LLVM-style inline assembly is unsupported in naked functions
    //~| WARN this was previously accepted
    core::hint::unreachable_unchecked();
}

#[naked]
unsafe extern "C" fn invalid_options() {
    asm!("", options(nomem, preserves_flags, noreturn));
    //~^ WARN asm options unsupported in naked functions: `nomem`, `preserves_flags`
    //~| WARN this was previously accepted
}

#[naked]
unsafe extern "C" fn invalid_options_continued() {
    asm!("", options(readonly, nostack), options(pure));
    //~^ ERROR asm with the `pure` option must have at least one output
    //~| WARN asm options unsupported in naked functions: `nostack`, `pure`, `readonly`
    //~| WARN this was previously accepted
    //~| WARN asm in naked functions must use `noreturn` option
    //~| WARN this was previously accepted
}

#[naked]
pub unsafe fn default_abi() {
    //~^ WARN Rust ABI is unsupported in naked functions
    asm!("", options(noreturn));
}

#[naked]
pub unsafe fn rust_abi() {
    //~^ WARN Rust ABI is unsupported in naked functions
    asm!("", options(noreturn));
}

#[naked]
pub extern "C" fn valid_a<T>() -> T {
    unsafe {
        asm!("", options(noreturn));
    }
}

#[naked]
pub extern "C" fn valid_b() {
    unsafe {
        {
            {
                asm!("", options(noreturn));
            };
        };
    }
}

#[naked]
pub unsafe extern "C" fn valid_c() {
    asm!("", options(noreturn));
}

#[cfg(target_arch = "x86_64")]
#[naked]
pub unsafe extern "C" fn valid_att_syntax() {
    asm!("", options(noreturn, att_syntax));
}

#[naked]
pub unsafe extern "C" fn inline_none() {
    asm!("", options(noreturn));
}

#[naked]
#[inline]
//~^ WARN naked functions cannot be inlined
//~| WARN this was previously accepted
pub unsafe extern "C" fn inline_hint() {
    asm!("", options(noreturn));
}

#[naked]
#[inline(always)]
//~^ WARN naked functions cannot be inlined
//~| WARN this was previously accepted
pub unsafe extern "C" fn inline_always() {
    asm!("", options(noreturn));
}

#[naked]
#[inline(never)]
//~^ WARN naked functions cannot be inlined
//~| WARN this was previously accepted
pub unsafe extern "C" fn inline_never() {
    asm!("", options(noreturn));
}

#[naked]
#[inline]
//~^ WARN naked functions cannot be inlined
//~| WARN this was previously accepted
#[inline(always)]
//~^ WARN naked functions cannot be inlined
//~| WARN this was previously accepted
#[inline(never)]
//~^ WARN naked functions cannot be inlined
//~| WARN this was previously accepted
pub unsafe extern "C" fn inline_all() {
    asm!("", options(noreturn));
}
