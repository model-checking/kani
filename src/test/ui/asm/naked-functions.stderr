error: asm with the `pure` option must have at least one output
  --> $DIR/naked-functions.rs:135:14
   |
LL |     asm!("", options(readonly, nostack), options(pure));
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^

error: patterns not allowed in naked function parameters
  --> $DIR/naked-functions.rs:22:5
   |
LL |     mut a: u32,
   |     ^^^^^

error: patterns not allowed in naked function parameters
  --> $DIR/naked-functions.rs:24:5
   |
LL |     &b: &i32,
   |     ^^

error: patterns not allowed in naked function parameters
  --> $DIR/naked-functions.rs:26:6
   |
LL |     (None | Some(_)): Option<std::ptr::NonNull<u8>>,
   |      ^^^^^^^^^^^^^^

error: patterns not allowed in naked function parameters
  --> $DIR/naked-functions.rs:28:5
   |
LL |     P { x, y }: P,
   |     ^^^^^^^^^^

error: referencing function parameters is not allowed in naked functions
  --> $DIR/naked-functions.rs:38:5
   |
LL |     a + 1
   |     ^
   |
   = help: follow the calling convention in asm block to use parameters

warning: naked functions must contain a single asm block
  --> $DIR/naked-functions.rs:35:1
   |
LL | / pub unsafe extern "C" fn inc(a: u32) -> u32 {
LL | |
LL | |
LL | |     a + 1
   | |     ----- non-asm is unsupported in naked functions
LL | |
LL | | }
   | |_^
   |
   = note: `#[warn(unsupported_naked_functions)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

error: referencing function parameters is not allowed in naked functions
  --> $DIR/naked-functions.rs:44:31
   |
LL |     asm!("/* {0} */", in(reg) a, options(noreturn));
   |                               ^
   |
   = help: follow the calling convention in asm block to use parameters

warning: only `const` and `sym` operands are supported in naked functions
  --> $DIR/naked-functions.rs:44:23
   |
LL |     asm!("/* {0} */", in(reg) a, options(noreturn));
   |                       ^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: naked functions must contain a single asm block
  --> $DIR/naked-functions.rs:51:1
   |
LL | / pub unsafe extern "C" fn inc_closure(a: u32) -> u32 {
LL | |
LL | |
LL | |     (|| a + 1)()
   | |     ------------ non-asm is unsupported in naked functions
LL | | }
   | |_^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: only `const` and `sym` operands are supported in naked functions
  --> $DIR/naked-functions.rs:71:10
   |
LL |          in(reg) a,
   |          ^^^^^^^^^
...
LL |          inlateout(reg) b,
   |          ^^^^^^^^^^^^^^^^
LL |          inout(reg) c,
   |          ^^^^^^^^^^^^
LL |          lateout(reg) d,
   |          ^^^^^^^^^^^^^^
LL |          out(reg) e,
   |          ^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: asm in naked functions must use `noreturn` option
  --> $DIR/naked-functions.rs:68:5
   |
LL | /     asm!("/* {0} {1} {2} {3} {4} {5} {6} */",
LL | |
LL | |
LL | |          in(reg) a,
...  |
LL | |          sym G,
LL | |     );
   | |_____^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: naked functions must contain a single asm block
  --> $DIR/naked-functions.rs:58:1
   |
LL | / pub unsafe extern "C" fn unsupported_operands() {
LL | |
LL | |
LL | |     let mut a = 0usize;
   | |     ------------------- non-asm is unsupported in naked functions
LL | |     let mut b = 0usize;
   | |     ------------------- non-asm is unsupported in naked functions
LL | |     let mut c = 0usize;
   | |     ------------------- non-asm is unsupported in naked functions
LL | |     let mut d = 0usize;
   | |     ------------------- non-asm is unsupported in naked functions
LL | |     let mut e = 0usize;
   | |     ------------------- non-asm is unsupported in naked functions
...  |
LL | |     );
LL | | }
   | |_^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: naked functions must contain a single asm block
  --> $DIR/naked-functions.rs:84:1
   |
LL | / pub extern "C" fn missing_assembly() {
LL | |
LL | |
LL | | }
   | |_^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: asm in naked functions must use `noreturn` option
  --> $DIR/naked-functions.rs:93:5
   |
LL |     asm!("");
   |     ^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: asm in naked functions must use `noreturn` option
  --> $DIR/naked-functions.rs:96:5
   |
LL |     asm!("");
   |     ^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: asm in naked functions must use `noreturn` option
  --> $DIR/naked-functions.rs:99:5
   |
LL |     asm!("");
   |     ^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: naked functions must contain a single asm block
  --> $DIR/naked-functions.rs:90:1
   |
LL | / pub extern "C" fn too_many_asm_blocks() {
LL | |
LL | |
LL | |     asm!("");
...  |
LL | |     asm!("");
   | |     -------- multiple asm blocks are unsupported in naked functions
...  |
LL | |     asm!("");
   | |     -------- multiple asm blocks are unsupported in naked functions
...  |
LL | |     asm!("", options(noreturn));
   | |     --------------------------- multiple asm blocks are unsupported in naked functions
LL | | }
   | |_^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

error: referencing function parameters is not allowed in naked functions
  --> $DIR/naked-functions.rs:110:11
   |
LL |         *&y
   |           ^
   |
   = help: follow the calling convention in asm block to use parameters

warning: naked functions must contain a single asm block
  --> $DIR/naked-functions.rs:107:5
   |
LL | /     pub extern "C" fn inner(y: usize) -> usize {
LL | |
LL | |
LL | |         *&y
   | |         --- non-asm is unsupported in naked functions
LL | |
LL | |     }
   | |_____^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: the LLVM-style inline assembly is unsupported in naked functions
  --> $DIR/naked-functions.rs:120:5
   |
LL |     llvm_asm!("");
   |     ^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>
   = help: use the new asm! syntax specified in RFC 2873
   = note: this warning originates in the macro `llvm_asm` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: naked functions must contain a single asm block
  --> $DIR/naked-functions.rs:117:1
   |
LL | / unsafe extern "C" fn llvm() -> ! {
LL | |
LL | |
LL | |     llvm_asm!("");
...  |
LL | |     core::hint::unreachable_unchecked();
   | |     ------------------------------------ non-asm is unsupported in naked functions
LL | | }
   | |_^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: asm options unsupported in naked functions: `nomem`, `preserves_flags`
  --> $DIR/naked-functions.rs:128:5
   |
LL |     asm!("", options(nomem, preserves_flags, noreturn));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: asm options unsupported in naked functions: `nostack`, `pure`, `readonly`
  --> $DIR/naked-functions.rs:135:5
   |
LL |     asm!("", options(readonly, nostack), options(pure));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: asm in naked functions must use `noreturn` option
  --> $DIR/naked-functions.rs:135:5
   |
LL |     asm!("", options(readonly, nostack), options(pure));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: Rust ABI is unsupported in naked functions
  --> $DIR/naked-functions.rs:144:15
   |
LL | pub unsafe fn default_abi() {
   |               ^^^^^^^^^^^
   |
   = note: `#[warn(undefined_naked_function_abi)]` on by default

warning: Rust ABI is unsupported in naked functions
  --> $DIR/naked-functions.rs:150:15
   |
LL | pub unsafe fn rust_abi() {
   |               ^^^^^^^^

warning: naked functions cannot be inlined
  --> $DIR/naked-functions.rs:190:1
   |
LL | #[inline]
   | ^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: naked functions cannot be inlined
  --> $DIR/naked-functions.rs:198:1
   |
LL | #[inline(always)]
   | ^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: naked functions cannot be inlined
  --> $DIR/naked-functions.rs:206:1
   |
LL | #[inline(never)]
   | ^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: naked functions cannot be inlined
  --> $DIR/naked-functions.rs:214:1
   |
LL | #[inline]
   | ^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: naked functions cannot be inlined
  --> $DIR/naked-functions.rs:217:1
   |
LL | #[inline(always)]
   | ^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

warning: naked functions cannot be inlined
  --> $DIR/naked-functions.rs:220:1
   |
LL | #[inline(never)]
   | ^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #32408 <https://github.com/rust-lang/rust/issues/32408>

error: aborting due to 8 previous errors; 25 warnings emitted

